src/cpp/common/tls_credentials_options_util.cc

```diff
@@ -0,0 +1,138 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+#include "src/cpp/common/tls_credentials_options_util.h"
+#include <grpcpp/security/tls_credentials_options.h>
+
+namespace grpc_impl {
+namespace experimental {
+
+/** Converts the Cpp key materials to C key materials; this allocates memory for
+ * the C key materials. Note that the user must free
+ * the underlying pointer to private key and cert chain duplicates; they are not
+ * freed when the UniquePtr<char> member variables of PemKeyCertPair are unused.
+ * Similarly, the user must free the underlying pointer to c_pem_root_certs. **/
+grpc_tls_key_materials_config* ConvertToCKeyMaterialsConfig(
+    const std::shared_ptr<TlsKeyMaterialsConfig>& config) {
+  if (config == nullptr) {
+    return nullptr;
+  }
+  grpc_tls_key_materials_config* c_config =
+      grpc_tls_key_materials_config_create();
+  ::grpc_core::InlinedVector<::grpc_core::PemKeyCertPair, 1>
+      c_pem_key_cert_pair_list;
+  for (auto key_cert_pair = config->pem_key_cert_pair_list().begin();
+       key_cert_pair != config->pem_key_cert_pair_list().end();
+       key_cert_pair++) {
+    grpc_ssl_pem_key_cert_pair* ssl_pair =
+        (grpc_ssl_pem_key_cert_pair*)gpr_malloc(
+            sizeof(grpc_ssl_pem_key_cert_pair));
+    ssl_pair->private_key = gpr_strdup(key_cert_pair->private_key.c_str());
+    ssl_pair->cert_chain = gpr_strdup(key_cert_pair->cert_chain.c_str());
+    ::grpc_core::PemKeyCertPair c_pem_key_cert_pair =
+        ::grpc_core::PemKeyCertPair(ssl_pair);
+    c_pem_key_cert_pair_list.push_back(::std::move(c_pem_key_cert_pair));
+  }
+  ::grpc_core::UniquePtr<char> c_pem_root_certs(
+      gpr_strdup(config->pem_root_certs().c_str()));
+  c_config->set_key_materials(std::move(c_pem_root_certs),
+                              std::move(c_pem_key_cert_pair_list));
+  c_config->set_version(config->version());
+  return c_config;
+}
+
+/** The C schedule and cancel functions for the credential reload config.
+ * They populate a C credential reload arg with the result of a C++ credential
+ * reload schedule/cancel API. **/
+int TlsCredentialReloadConfigCSchedule(void* config_user_data,
+                                       grpc_tls_credential_reload_arg* arg) {
+  if (arg == nullptr || arg->config == nullptr ||
+      arg->config->context() == nullptr) {
+    gpr_log(GPR_ERROR, "credential reload arg was not properly initialized");
+    return 1;
+  }
+  TlsCredentialReloadConfig* cpp_config =
+      static_cast<TlsCredentialReloadConfig*>(arg->config->context());
+  TlsCredentialReloadArg* cpp_arg = new TlsCredentialReloadArg(arg);
+  int schedule_result = cpp_config->Schedule(cpp_arg);
+  delete cpp_arg;
```

Comment:
I think you cannot delete it here. Otherwise, when cancelling with it, users can touch into deleted memory.

Reply:
Thanks for your comment @yang-g. Following a discussion with @yihuazhang, I removed this "delete" line, and added a `destroy_context` function pointer to the definition of `grpc_tls_credential_reload_arg`, and this is populated by the function `TlsCredentialReloadArgDestroyContext` in the util file. This `destroy_context` method is called immediately prior to deleting the C arg, and hence it deletes the associated C++ arg immediately before deleting the C arg.